src/devices/Card/Ndef/NdefMessage.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+
+namespace Iot.Device.Ndef
+{
+    /// <summary>
+    /// Create a NDEF = NFC Data Exchange Format Message class
+    /// </summary>
+    public class NdefMessage
+    {
+        /// <summary>
+        /// Associated with the GeneralPurposeByteConsitions, it tells if a sector is read/write and a valid
+        /// NDEF sector
+        /// </summary>
+        public const byte GeneralPurposeByteNdefVersion = 0b0100_0000;
+
+        /// <summary>
+        /// List of all NDEF Records
+        /// </summary>
+        public List<NdefRecord> Records { get; set; } = new List<NdefRecord>();
+
+        /// <summary>
+        /// Create an empty NDEF Message
+        /// </summary>
+        public NdefMessage()
+        {
+        }
+
+        /// <summary>
+        /// Create NDEF Message from a span of bytes
+        /// </summary>
+        /// <param name="message">the message in span of bytes</param>
+        public NdefMessage(ReadOnlySpan<byte> message)
+        {
+            int idxMessage = 0;
+            while (idxMessage < message.Length)
+            {
+                var ndefrec = new NdefRecord(message.Slice(idxMessage));
+                Records.Add(ndefrec);
+                idxMessage += ndefrec.Length;
+            }
+        }
+
+        /// <summary>
+        /// Get the length of the message
+        /// </summary>
+        public int Length => Records.Select(m => m.Length).Sum();
+
+        /// <summary>
+        /// Serialize the message in a span of bytes
+        /// </summary>
+        /// <param name="messageSerialized">Span of bytes for the serialized message</param>
+        public void Serialize(Span<byte> messageSerialized)
+        {
+            if (messageSerialized.Length < Length)
+            {
+                throw new ArgumentException($"Span of bytes needs to be at least as large as the Message total length");
+            }
+
+            if (Records.Count == 0)
+            {
+                return;
+            }
+
+            // Make sure we set correctly the Begin and End message flags
+            Records.First().Header.MessageFlag |= MessageFlag.MessageBegin;
+            Records.First().Header.MessageFlag &= ~MessageFlag.MessageEnd;
+            Records.Last().Header.MessageFlag |= MessageFlag.MessageEnd;
+
+            int idx = 0;
```

Comment:
instead of tracking index it's usually better idea to slice the span directly so that you don't accidentally write somewhere in between

Reply:
I am slicing the span but that would need every time to keep a Span all the time and reduce it all the time.